staging/src/k8s.io/client-go/tools/clientcmd/validation.go

```diff
@@ -185,9 +185,10 @@ func validateClusterInfo(clusterName string, clusterInfo clientcmdapi.Cluster) [
 	}
 	if len(clusterInfo.CertificateAuthority) != 0 {
 		clientCertCA, err := os.Open(clusterInfo.CertificateAuthority)
-		defer clientCertCA.Close()
```

Comment:
this is bad style. If anything goes wrong below before the close, we leak a file.

Reply:
also below

Reply:
How about defer after check returned error?  
like this:
```go
		clientCertCA, err := os.Open(clusterInfo.CertificateAuthority)
		if err != nil {
			validationErrors = append(validationErrors, fmt.Errorf("unable to read certificate-authority %v for %v due to %v", clusterInfo.CertificateAuthority, clusterName, err))
		}
		defer clientCertCA.Close()
```

Otherwise, we need to ignore following check if we back to the way it was.
`
vendor/k8s.io/client-go/tools/clientcmd/validation.go:188:3: should check returned error before deferring clientCertCA.Close() (SA5001)
`

Reply:
Yeah, you are right. I'd prefer to ignore the noisy from staticcheck.

Reply:
> //lint:ignore SA5001 defer file close immediately to avoid resource(file) leak

That is incorrect. Generally speaking, if os.Open returns an error, then the value of the returned file is undefined, and calling methods on it may behave incorrectly or panic. The user isn't supposed to close the file, and there is nothing to leak. Concretely speaking, if os.Open returns an error, it returns a nil file, so there's concretely nothing to leak, either.

Now, calling (*os.File).Close on a nil file does "work", in that Close explicitly checks whether the file is nil, so that it can return early and not crash. I would consider that a convenience. It's not a documented feature, at any rate. And it does cause Close to return an error.

Also note that "it doesn't crash" only applies to os.Open. Any other function that opens a resource and also returns an error may very well crash if trying to close the resource that is in an undefined state.

The canonical pattern very much is

```
f, err := os.Open(...)
if err != nil { ... }
defer f.Close()
```

Reply:
I am fine that that style of `defer` after error handling.